Step1 문자열 덧셈 계산기

jaeykweon opened the pull request:
안녕하세요 현구막님
문자열 덧셈 계산기 리뷰요청 드립니다

이전 리뷰를 적용하였으며,
이번 과제는 TDD 방식으로 구현해 보았습니다

추가적으로 드리고 싶은 질문은 확장성을 어디까지 고려해야하는 지입니다.
지난 자동차 경주의 경우 어떻게 변해갈지 예상이 어려웠다면, 이번에는 계산기라는 틀이 정해져 있습니다.

이때 주어진 조건만을 충족하는 최소한의 코드를 작성해야할지
아니면 나중에 사칙연산 및 고급 연산(루트 등), 소괄호를 활용한 우선순위 연산까지 고려하여 
미리 중위연산 까지 대비해야할지가 고민이었습니다

현구막님께서는 어떤 인사이트를 가지고 계신지 궁금합니다
감사합니다

jaeykweon committed:
step0 코드리뷰 적용

jaeykweon committed:
feat: 첫번째 요구사항 충족

jaeykweon committed:
feat: 두번째 요구사항 충족

jaeykweon committed:
feat: 세번째 요구사항 충족

jaeykweon committed:
feat: 네번째 요구사항 충족

jaeykweon committed:
feat: 다섯번째 요구사항 충족

jaeykweon committed:
feat: 여섯번째 요구사항 충족

jaeykweon committed:
refactor: 추가적인 음수 케이스를 위한 리팩토링

Hyeon9mak commented on src/main/java/calculator/NegativeNotAllowedException.java:
```diff
@@ -0,0 +1,8 @@
+package calculator;
+
+public class NegativeNotAllowedException extends RuntimeException {
+
+    public NegativeNotAllowedException() {
+        super("음수는 입력할 수 없습니다.");
+    }
+}
```
기본적으로 제공되는 `IllegalArgumentsException` 을 활용하지 않고 custom exception 을 만들어서 메세지를 고정하신 이유가 궁금해요.

jaeykweon replied:
저는 보편적인 입력값 오류에 대해서는 IllegalArgumentException 을 사용합니다
(ex. 숫자가 들어와야하는데 문자가 들어오는 경우, null이면 안되는데 null이 들어온 경우)
위 에러는 비지니스적으로 의미있는  오류라고 판단하여 따로 커스텀 익셉션을 추가하였습니다

Hyeon9mak replied:
비즈니스 오류에 대해 custom exception 을 사용하시는 이유가 궁금합니다!
이를 통해 편의를 느꼈던 사례나, 기본으로 제공되는 exception 을 이용했을 때 불편했던 경험이 있으신걸까요?

Hyeon9mak commented on src/main/java/calculator/PositiveNumber.java:
```diff
@@ -0,0 +1,18 @@
+
+    public PositiveNumber add(PositiveNumber number) {
+        return new PositiveNumber(value + number.value);
+    }
```
편의를 위해 덧셈 연산 메서드를 만들어주셨군요 👍

Hyeon9mak commented on src/main/java/calculator/PositiveNumber.java:
```diff
@@ -0,0 +1,18 @@
+        if (value < 0) {
+            throw new NegativeNotAllowedException();
+        }
+    }
```
앞으로 `PositiveNumber` vo 가 만들어지면 믿고 사용할 수 있겠네요!
별건 아니지만... 0 은 양수가 아니긴하죠 😄

jaeykweon replied:
이 부분도 놓쳤네요 NotNegativeNumber로 변경하였습니다!

Hyeon9mak commented on src/main/java/calculator/Separator.java:
```diff
@@ -0,0 +1,30 @@
+    private final List<String> separators = new ArrayList<>(Arrays.asList(",", ":"));
+
+    public Stream<PositiveNumber> separate(String input) {
+        if (input.startsWith("//")) {
+            String[] split = input.split("\n");
+            String delimiter = split[0].substring(2);
+            String cleaned = split[1];
+            separators.add(delimiter);
+            return separateAll(cleaned);
+        }
```
분해한 것의 0번째 인덱스는 무얼 찾아내기 위함이고,
거기서 2글자만 발라내는 이유는 무엇일까요?

이 과정을 따로 메서드로 구분하여 이름을 붙여주거나, 인덱스 값을 상수로 구분하여 이름을 붙여주면
동료 개발자가 코드를 읽을 때 훨씬 수월하지 않을까요? 😄 👍

jaeykweon replied:
Pattern 매칭으로 변경하여 위 코드는 삭제되었습니다
좋은 리뷰 감사합니다!

Hyeon9mak commented on src/test/java/calculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,63 @@
+        stringCalculator = new StringCalculator();
+    }
+
+    @DisplayName(value = "빈 문자열 또는 null 값을 입력할 경우 0을 반환해야 한다.")
+    @ParameterizedTest
+    @NullAndEmptySource
+    public void emptyOrNull(final String text) {
+        assertThat(stringCalculator.add(text)).isZero();
+    }
```
`NullAndEmptySource` 활용 👍 

혹시 구분자 하나만 전달되면 어떤 일이 벌어질까요?

jaeykweon replied:
그 부분은 미쳐 생각못했습니다 테스트에 추가하였습니다
`구분자만 있는 경우 0을 반환한다` 테스트를 확인해주시면 됩니다

Hyeon9mak reviewed (changes requested):
안녕하세요 재용님~ 설연휴 잘 보내셨나요?
1단계도 잘 진행해주셨습니다 👍 

> 확장성을 어디까지 고려해야하는 지입니다.

> 주어진 조건만을 충족하는 최소한의 코드를 작성해야할지
아니면 나중에 사칙연산 및 고급 연산(루트 등), 소괄호를 활용한 우선순위 연산까지 고려하여
미리 중위연산 까지 대비해야할지가 고민이었습니다.

재용님도 잘 아시겠지만 미리 고려하고 대비 할수록 나중이 편해지고,
주어진 조건만 충족하면 지금이 편해지죠.
"그 사이에서 줄타기가 중요하다." 와 같은 재미없는 이야기는 저도 싫어하는데요,

결국 줄을 탈 수 있는 기준이 되는 건 "내가 이 도메인에 대해 얼마나 잘 알고 있느냐" 라고 생각합니다.
도메인에 대해 잘 알면 알수록, 앞으로 어떤 부분에 변경이 생길지 쉽게 예측이 가능하기 때문이에요.
가령 "문자열 덧셈기에는 앞으로 또 다른 구분자가 추가되는 일이 잦다." 등이 있겠네요!

문자열 덧셈 계산기를 현실적으로 생각해보면 어떨까요?
재용님이 앞으로 문자열 덧셈 계산기 코드를 유지보수하실 일이 생길까요?
당연히 없을테고, 그럼 너무 힘을 들이실 필요가 없겠네요!

...만 미션을 이렇게 진행하면 너무 재미가 없으니 😄 
재용님께서 특정한 미래 상황을 상상하시고, 코드를 작성하신 다음
왜 이렇게 작성했는지 리뷰어에게 상황을 설명해주시면 서로 즐거운 미션이 될거 같습니다 👍 

간단한 코멘트 드렸으니 확인해주세요.

jaeykweon committed:
refactor: 코드 리뷰 적용
- PositiveNumber -> NotNegativeNumber 네이밍 변경
- 계산기 초기값 0으로 지정
- 구분자만 넣는 경우 테스트 추가
- 커스텀 구분자 탐색 방식을 Pattern과 Matcher 로 변경
- Matcher의 동작 방식을 보여주는 테스트 클래스 구성

Hyeon9mak commented on src/main/java/calculator/Separator.java:
```diff
@@ -0,0 +1,37 @@
+
+    private final List<String> separators = new ArrayList<>(Arrays.asList(",", ":"));
+
+    public Stream<NotNegativeNumber> separate(String input) {
+        Pattern pattern = Pattern.compile("//(.*?)\\n");
+        Matcher matcher = pattern.matcher(input);
+
```
매 연산마다 `Pattern` 인스턴스가 생성되겠네요. 이 역시도 상수로 다뤄보는 건 어떨까요? 😉

Hyeon9mak commented on src/test/java/calculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,70 @@
+                .isThrownBy(() -> stringCalculator.add(text));
+    }
+
+    @DisplayName(value = "구분자만 있는 경우 0을 반환한다.")
+    @ParameterizedTest
+    @ValueSource(strings = {":"})
+    void onlyDelimiter(final String text) {
+        assertThat(stringCalculator.add(text)).isEqualTo(0);
+    }
```
😉 👍

Hyeon9mak reviewed (approved):
안녕하세요 재용님! 피드백 잘 반영해주셨습니다.
간단한 질문 코멘트 하나 더 남겼으니 확인해주세요!
다음 단계 진행과 병행해주시면 될거 같습니다 😄
